### PR TITLE
[WHIT-3696] Access limiting assets

### PR DESCRIPTION
**Pull request by lauraghiorghisor-tw**
## What & why

These changes ensure that access limitation is set on assets of `AttachmentData`s (the "assetable"), from organisations- and individuals-based access limitation at the edition (the "attachable") level.

Changes:
- Set the feature flags to true in integration so we can test the changes. Why: we use the flags in models, and the controller flipflop value does not reach there.
- Set asset access limitation to organisations when flag is on -> mainly required test coverage, as we tap into the pre-existing infrastructure for organisations; changes required for response-type attachables.
- Set asset access limitation to individuals when flag is on -> required changes to the creation and update steps in the asset lifecycle, including custom changes for response-type attachables.
- Add update service listeners to the "admin" access limiting controller, so that changes to access limitation also update assets downstream.
- Refactor the create job tests to make them clearer/more explicit. This is in preparation for future changes that might fix the responses "bug" identified.

### Feature flags

- Flag default on in integration
- Flag NOT default on in staging

### Assets 

All editions below had assets - the permissions were checked against the assets.
- ✅ Remove and add AL organisations via the edition form - config driven + legacy editions
- ✅ Remove and add individuals via the edition form - config driven + legacy editions
- ✅ Change org access from admin view - config driven + legacy editions [found bug, now fixed in PR]
- ✅ Change individuals access from admin view - config driven + legacy [🐛bug found, see note above]
- ✅ Access limitation is removed on publish - config driven + legacy editions
- ✅ Asset update (e.g. attachment title change) -> maintains AL (tested orgs only)
- ✅ Asset Delete - no-op
- ✅ Asset Replacement -> redirects and new assets has AL (tested orgs only)
- ✅ Check some asset types that instantly download (ods) - AL works as expected for them too
- ✅ Check other types of attachables 
    - ✅ Consultation responses OK [bar the issue 🐞 detailed above]
    - 🔧 Cons response form data - don't work yet, work in progress on a different card
- ✅ Preview link -  still allows sidestepping the restrictions (for edition and assets)
     - 🤔 Once the cookie is set, I can access the pages even if the permissions is removed - which is probably expected, but I wonder when the cookies will expire.

## Bugs found

- 🚨 Removing individual emails and providing valid, but not "real" signon email instead, saves ok, blocks permissions in WH to removed emails, but does not send it downstream -> preview in draft is still accessible, and so are assets. If the replacement input contains at least one valid email, then it works.

[Jira](https://gov-uk.atlassian.net/browse/WHIT-3696)

### CODE REVIEW

File: app/controllers/admin/edition_access_limited_controller.rb [hunk 1]
```diff
@@ -14,8 +14,9 @@ def update
       return render :edit
     end
 
-    if @edition.save
-      PublishingApiDocumentRepublishingJob.perform_async(@edition.document_id, false)
+    if updater.can_perform? && @edition.save
```
Comment (lauraghiorghisor-tw) [inline app/controllers/admin/edition_access_limited_controller.rb:17]: Kind of in two minds about this - we can also just call the republish job and the attachment updater here. I don't _think_ it has any weird consequences. It does provide the security of tapping into the service listeners for all future changes. The original issue I was solving here was that the asset updates were not happening; with this change we have that for free.

I did test some regular changes to orgs, not just access limiting orgs, seems ok, goes downstream, nothing catches fire.

File: app/models/consultation_response.rb [hunk 2]
```diff
@@ -19,6 +19,10 @@ def access_limited_object
 
   delegate :organisations, to: :parent_attachable
 
+  delegate :access_limiting_organisations?, to: :parent_attachable
```
Comment (lauraghiorghisor-tw) [inline app/models/consultation_response.rb:22]: I think there is another way if we don't want to do this, i.e. using the same call in the create job, as we do for updater (see commit msg). Will have a different PR for that, as it needs further testing. Either way, this meets the need for now.

File: app/models/consultation_response.rb [hunk 3]
```diff
@@ -19,6 +19,10 @@ def access_limited_object
 
   delegate :organisations, to: :parent_attachable
```
Comment (lauraghiorghisor-tw) [inline app/models/consultation_response.rb:20]: I imagine this can be removed once the flags are on.

File: config/features.rb [hunk 4]
```diff
@@ -35,8 +35,8 @@
           default: Rails.env.development?
   feature :access_limiting_organisations_ui,
           description: "Replace the access-limiting checkbox with radio + organisation selector UI",
-          default: false
+          default: Whitehall.integration? || false
   feature :access_limiting_individuals_ui,
           description: "Extend the access-limiting radio UI with an individuals option",
-          default: false
+          default: Whitehall.integration? || false
```
Comment (jamiestamp) [inline config/features.rb]: We can drop ` || false` here as `.integration?` is returning a boolean

File: test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb
Comment (jamiestamp) [inline test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb:1]: Is it worth here having a follow up task to bring the other tests inline with the new approach here? It feels like if it was problematic on other tests, then we would benefit from updating the others to have more confidence in the tests.

There's also a lot of repeated scaffolding, i.e.:

```ruby
asset_params = {
  assetable_id: assetable.id,
  asset_variant: Asset.variants[:original],
  assetable_type: assetable.class.to_s,
}.deep_stringify_keys
```
So a helper would help tidy up the code a bit here.

This isn't a blocker on approving, but would be good to tackle soon.

File: app/models/attachment_data.rb [hunk 5]
```diff
@@ -149,10 +149,16 @@ def attachable_url
...
     return [] unless access_limited?
 
-    AssetManagerAccessLimitation.for(access_limited_object)
+    AssetManagerAccessLimitation.for_organisations(access_limited_object)
+  end
+
+  def access_limited_individual_ids
+    return [] unless access_limited?
+
+    AssetManagerAccessLimitation.for_individuals(access_limited_object)
```
Comment (jamiestamp) [inline app/models/attachment_data.rb]: These could be made consistent in how they are named. Perhaps opt for the `access_limitation_` prefix to match what we had before

File: test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb
Comment (lauraghiorghisor-tw) [reply test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb:1]: I think the non-AL tests are kind of generic, you don't create new files for what we are testing, so they can just use the assetable set up at the top. I am happy to keep the verbose setup in the AL context only.

The AL tests don't REALLY need the custom assetable either - because we still only use the attachable value in the code to deduce the AL. It WOULD have mattered if I implemented the thing I did not implement after all - using the assetable rather than the attachable for access limitation (went for delegation instead because I was scared I might break other things - but I have it on a branch for when I have a bit more brain). But I chose to keep the test refactor because it just felt wrong to set up this new file attachment with a different assetable, while using the unrelated assetable from the setup in the test. It's just wrong and it makes you wonder when you have a failing tests - is it that, is it the other thing? And some of the tests for responses were completely wrong before especially around responses!

I am at a point in life where I refuse to DRY anything especially in this codebase 😅 I want to read it top down, and not have to jump into any other methods to understand what it does. So am happy with a bit of repetition if it doesn't take me out of the flow of the test/code. But I can consider adding a helper method here, fine! 😅

File: app/models/attachment_data.rb (hunk 5 above)
Comment (lauraghiorghisor-tw) [reply app/models/attachment_data.rb]: You named that one! (I think?) 😝 I honestly did not even spot that once in all the 57 rebases and edits I must have done on this PR.

Comment (lauraghiorghisor-tw) [reply app/models/attachment_data.rb]: I can rename these two to be consistent - following the original org name. But that is NOT like how everything else is named. But that is ok, within this class things are "limitation" rather than "limiting", I can live with that.

File: app/controllers/admin/edition_access_limited_controller.rb (hunk 1 above)
Comment (TonyGDS) [reply app/controllers/admin/edition_access_limited_controller.rb:17]: There is an edge case. If you are on this screen and the edition is published at that moment, when you make changes and click save there is no error message. You just stay on the page. Also there is no test for this scenario.

File: test/unit/lib/whitehall/authority/standard_edition_rules_test.rb
Comment (TonyGDS) [inline test/unit/lib/whitehall/authority/standard_edition_rules_test.rb:154]: This has been improved since the first commit so not a big issue but maybe we could group these tests under contexts for the feature flag e.g. `access_limiting_organisations_ui flag is ON`. You can still keep the full setup in the tests so we aren't DRYing anything and the duplication etc so it reads top down, but we don't need to add "flag is ON/OFF" to each test. When it comes to removing the flags after the release it's easier to clean up the tests.

File: test/unit/app/services/asset_manager/attachment_updater_test.rb [hunk 6]
```diff
@@ -64,6 +64,31 @@ class AssetManager::AttachmentUpdaterTest < ActiveSupport::TestCase
 
         AssetManager::AttachmentUpdater.call(attachment.attachment_data)
       end
+
+      it "sets the expected asset attributes when the access_limiting_organisations_ui flag is on" do
```
Comment (TonyGDS) [inline test/unit/app/services/asset_manager/attachment_updater_test.rb:71]: can we be more specific here about what the expectation is? Otherwise just say "sets the asset attributes"

File: test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb [hunk 7]
```diff
@@ -47,25 +47,48 @@ class AssetManagerCreateAssetJobTest < ActiveSupport::TestCase
...
-    attachment.attachment_data.attachable = consultation
+    FactoryBot.create(:file_attachment, attachable: consultation)
 
     Services.asset_manager.expects(:create_asset).with(has_entry(access_limited_organisation_ids: [@organisation.content_id])).returns(@asset_manager_response)
 
     @job.perform(@file.path, @asset_params, true, consultation.class.to_s, consultation.id)
   end
 
-  test "marks attachments belonging to consultation responses as access limited" do
+  test "marks attachments belonging to consultations as access limited to organisations, when flag is on" do
```
Comment (TonyGDS) [inline test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb]: maybe be can group these "flag on" tests under a context, so it's easier to remove them later

File: test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb
Comment (TonyGDS) [inline test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb:1]: Should we not add a tests here for call for evidence outcomes as well? If not, can we make the test names more generic i.e. not referring to the specific types

File: test/functional/admin/edition_access_limited_controller_test.rb [hunk 8]
```diff
@@ -497,4 +495,62 @@ class Admin::EditionAccessLimitedControllerTest < ActionController::TestCase
     assert edition.reload.access_limiting_individuals.exists?(email: "user@example.com")
     assert_not edition.access_limiting_individuals.exists?(email: "another_user@example.com")
   end
+
+  # flags agnostic
```
Comment (TonyGDS) [inline test/functional/admin/edition_access_limited_controller_test.rb:499]: need to make sure we remove these later during clean up, or remove now

File: app/sidekiq/asset_manager_create_asset_job.rb [hunk 9]
```diff
@@ -21,7 +21,9 @@ def perform(temporary_location, asset_params, draft = false, attachable_model_cl
 
     asset_options = { file:, auth_bypass_ids:, draft: }
     authorised_organisation_uids = get_authorised_organisation_ids(attachable_model_class, attachable_model_id)
-    asset_options[:access_limited_organisation_ids] = authorised_organisation_uids if authorised_organisation_uids
+    asset_options[:access_limited_organisation_ids] = authorised_organisation_uids if authorised_organisation_uids&.any?
```
Comment (TonyGDS) [inline app/sidekiq/asset_manager_create_asset_job.rb]: This no longer sets the key if authorised_organisation_uids, is an empty array. That's fine if no downstream impact.

File: app/controllers/admin/edition_access_limited_controller.rb (hunk 1 above)
Comment (lauraghiorghisor-tw) [reply app/controllers/admin/edition_access_limited_controller.rb:17]: I wonder what it did before my change - no one should be altering AL for published editions so that should not even be a case. Either way, that updater can perform might return an error, I just need to surface it on the form, now that I added it.

File: test/unit/lib/whitehall/authority/standard_edition_rules_test.rb
Comment (lauraghiorghisor-tw) [reply test/unit/lib/whitehall/authority/standard_edition_rules_test.rb:154]: Not my change, think PR needs a rebase. Won't do.

File: test/functional/admin/edition_access_limited_controller_test.rb (hunk 8 above)
Comment (lauraghiorghisor-tw) [reply test/functional/admin/edition_access_limited_controller_test.rb:499]: The reason they were added was to help with cleanup - I imagine we will also clean them up when we clean up.

File: app/controllers/admin/edition_access_limited_controller.rb (hunk 1 above)
Comment (lauraghiorghisor-tw) [reply app/controllers/admin/edition_access_limited_controller.rb:17]: image

Before my change it worked 😅 I think it now exits early on the first half if the `updater.can_perform? && @edition.save` because the updater cannot perform, and we never call `edition.save` so when we render the edit template there are no errors on the edition. On the regular edition controller we render the updater's failure reason.

Good catch 💕

File: test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb (hunk 7 above)
Comment (lauraghiorghisor-tw) [reply test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb]: I'd say they're obvious enough, and not too many. I can live with it.

File: test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb
Comment (lauraghiorghisor-tw) [reply test/unit/app/sidekiq/asset_manager_create_asset_job_test.rb:1]: Done.

File: app/controllers/admin/edition_access_limited_controller.rb (hunk 1 above)
Comment (lauraghiorghisor-tw) [reply app/controllers/admin/edition_access_limited_controller.rb:17]: image


All done.

File: app/sidekiq/asset_manager_create_asset_job.rb (hunk 9 above)
Comment (lauraghiorghisor-tw) [reply app/sidekiq/asset_manager_create_asset_job.rb]: I think there is no major effect, because the updater then sets it all to `[]` anyway (I so want to remove that updater enqueueing from there) - but I suppose it makes no sense for the create to set it when it can just be nil. The updater will have to set it to [] because it needs to override previous settings so that makes sense. 

Pushed this commit https://github.com/alphagov/whitehall/pull/11601/changes/1d330aa91e06a23a1f5b3c177fc633299f0bec2e. I can fix it up in the commit that introduces individuals if you want.

File: test/unit/app/services/asset_manager/attachment_updater_test.rb (hunk 6 above)
Comment (lauraghiorghisor-tw) [reply test/unit/app/services/asset_manager/attachment_updater_test.rb:71]: Neither version is enlightening me - leaving as is, if that's ok?

File: app/sidekiq/asset_manager_create_asset_job.rb (hunk 9 above)
Comment (lauraghiorghisor-tw) [reply app/sidekiq/asset_manager_create_asset_job.rb]: Pushed again.

Decided to:
- not send access limitation at create time; send [] if genuinely returned by the payload builder.
- send [] on update when there is no access limitation or the right access limitation cannot be found

Asset Manager will error if we send nil as the field value for these fields, so we can only send something not nil or blank. See https://github.com/alphagov/asset-manager/blob/b282a2da6639dc9a5203d96a4f269c64deeba1b7/app/controllers/assets_controller.rb#L60 -> that empty check.

The documentation in asset manager does say we can send [] on create as well.

File: app/models/attachment_data.rb [hunk 10]
```diff
@@ -149,10 +149,16 @@ def attachable_url
...
-  def access_limitation
+  def access_limitation_organisation_ids
     return [] unless access_limited?
 
-    AssetManagerAccessLimitation.for(access_limited_object)
+    AssetManagerAccessLimitation.for(access_limited_object, :organisations) || []
+  end
+
+  def access_limitation_individual_ids
+    return [] unless access_limited?
```
Comment (lauraghiorghisor-tw) [inline app/models/attachment_data.rb:159]: These early returns here are unbreakable - cannot touch them 😅

I think it is worth - when we have some more time - looking into why that is needed, whether we even need an access limited object etc etc.

**Review by jamiestamp — APPROVED** [review]
